src/impl_/coroutine.rs

```diff
@@ -0,0 +1,36 @@
+use std::future::Future;
+
+use crate::{
+    coroutine::Coroutine,
+    types::{PyModule, PyString},
+    IntoPy, PyClass, PyErr, PyObject, Python,
+};
+
+pub fn new_coroutine<F, T, E>(name: &PyString, qualname: &PyString, future: F) -> Coroutine
+where
+    F: Future<Output = Result<T, E>> + Send + 'static,
+    T: IntoPy<PyObject>,
+    E: Into<PyErr>,
+{
+    Coroutine::new(Some(name.into()), Some(qualname.into()), future)
+}
+
+pub fn coroutine_qualname<'py>(
+    py: Python<'py>,
+    module: Option<&PyModule>,
+    name: &str,
+) -> &'py PyString {
+    match module.and_then(|m| m.name().ok()) {
+        Some(module) => PyString::new(py, &format!("{}.{}", module, name)),
+        None => PyString::new(py, name),
+    }
+}
+
+pub fn method_coroutine_qualname<'py, T: PyClass>(py: Python<'py>, name: &str) -> &'py PyString {
+    let class = T::NAME;
+    let qualname = match T::MODULE {
+        Some(module) => format!("{}.{}.{}", module, class, name),
+        None => format!("{}.{}", class, name),
+    };
+    PyString::new(py, &qualname)
+}
```

Comment:
I wonder, to avoid Python string creation with every coroutine, can this be done lazily (by e.g. storing a reference to `T::type_object`, or taking a function pointer, or storing a reference to the module etc).

Reply:
I've finally used a `qualname_prefix` instead of passing the `__qualname__` directly. 
The constructor parameter may be modified when it will be exposed (maybe using `Option<Cow<'static, str>>` instead of `Option<&'static str>` for example), but for now, it's adapted to the macro and that's enough.